src/checkpoint/checkpoint_txn.c

```diff
@@ -2782,3 +2817,316 @@ __checkpoint_timing_stress(WT_SESSION_IMPL *session, uint64_t flag, struct times
+  WT_DATA_HANDLE *handle, WT_TXN_SNAPSHOT *snapshot, const char *cfg[])
+{
+    WTI_CKPT_WORK_UNIT *entry;
+
+    WT_RET(__wt_calloc_one(session, &entry));
+    entry->config = cfg;
+    entry->snapshot = snapshot;
+    entry->handle = handle;
+
+    TAILQ_INSERT_TAIL(&queue->qh, entry, q);
+
+    return (0);
+}
+
+/*
+ * __checkpoint_push_dhandles --
+ *     Push all handles to the checkpoint threads for checkpoint operation.
+ */
+static int
+__checkpoint_push_dhandles(WT_SESSION_IMPL *session, const char *cfg[], int *dhandles_countp)
+{
+    WT_DECL_RET;
+    WTI_CKPT_WORKERS *workers = S2C(session)->ckpt.workers;
+
+    (*dhandles_countp) = 0;
+
+    /* If we have already locked the handles, apply the operation. */
+    __wt_spin_lock(session, &workers->to_process_queue.lock);
+    for (u_int i = 0; i < session->ckpt.handle_next; ++i) {
+        if (session->ckpt.handle[i] == NULL)
+            continue;
+        WT_ERR(__checkpoint_push_work_to_process(session, &workers->to_process_queue,
+          session->ckpt.handle[i], &session->txn->snapshot_data, cfg));
+        (*dhandles_countp)++;
+    }
+
+err:
+    __wt_spin_unlock(session, &workers->to_process_queue.lock);
+
+    /* If no errors reported, signal all the workers to start the job. */
+    if (ret == 0)
+        __wt_cond_signal(session, workers->to_process_queue.cond);
+
+    return (ret);
+}
+
+/*
+ * __checkpoint_push_work_result --
+ *     Push a work unit to the the main thread to update metadata.
+ */
+static int
+__checkpoint_push_result(WT_SESSION_IMPL *session, WT_DATA_HANDLE *dhandle, const char *ckptlsn_str)
+{
+    WTI_CKPT_WORK_UNIT *entry;
+
+    WTI_CKPT_SAFE_WORK_QUEUE *result_queue = &S2C(session)->ckpt.workers->result_queue;
+
+    WT_ASSERT(session, session->dhandle == dhandle);
+    WT_ASSERT(session, ckptlsn_str != NULL);
+    WT_RET(__wt_calloc_one(session, &entry));
+    entry->handle = dhandle;
+    memcpy(entry->lsn_str, ckptlsn_str, WT_MAX_LSN_STRING);
+
+    __wt_spin_lock(session, &result_queue->lock);
+    TAILQ_INSERT_TAIL(&result_queue->qh, entry, q);
+    __wt_spin_unlock(session, &result_queue->lock);
+    __wt_cond_signal(session, result_queue->cond);
+
+    return (0);
+}
+
+static int
+__checkpoint_result_meta_ckptlist_set(WT_SESSION_IMPL *session, int *dhandles_donep)
+{
+    WT_BTREE *btree;
+    WTI_CKPT_WORK_UNIT *entry;
+    WT_DECL_RET;
+
+    WTI_CKPT_SAFE_WORK_QUEUE *result_queue = &S2C(session)->ckpt.workers->result_queue;
+
+    __checkpoint_pop_work(session, result_queue, &entry);
+    if (entry == NULL)
+        return (0);
+
+    WT_ASSERT(session, entry->config == NULL);
+    WT_ASSERT(session, entry->snapshot == NULL);
+    WT_ASSERT(session, strlen(entry->lsn_str) != 0);
+    WT_ASSERT(session, entry->handle != NULL);
+
+    btree = entry->handle->handle;
+    WT_ERR(__wt_meta_ckptlist_set(session, entry->handle, btree->ckpt, entry->lsn_str));
```

Comment:
We need `WT_WITH_DHANDLE` macro used here.

Reply:
Do you mean something like:

```c
    WT_WITH_DHANDLE(session, entry->handle, WT_ERR(__wt_meta_ckptlist_set(session, entry->handle, btree->ckpt, entry->lsn_str)));
```

?

Reply:
Yes